mcnode/gc.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	sqlite3 "github.com/mattn/go-sqlite3"
+	mcq "github.com/mediachain/concat/mc/query"
+	pb "github.com/mediachain/concat/proto"
+	multihash "github.com/multiformats/go-multihash"
+)
+
+var (
+	NodeMustBeOffline = errors.New("Node must be offline")
+)
+
+func (node *Node) doGC(ctx context.Context) (int, error) {
+	if node.status != StatusOffline {
+		return 0, NodeMustBeOffline
+	}
+
+	gc := &GCDB{}
+	err := gc.Open(node.home)
+	if err != nil {
+		return 0, err
+	}
+	defer gc.Close()
+
+	err = gc.Merge(ctx, node.db)
+	if err != nil {
+		return 0, err
+	}
+
+	return gc.GC(ctx, node.ds)
+}
+
+func (node *Node) doCompact() error {
+	node.ds.Compact()
+	return nil
+}
+
+type GCDB struct {
+	db        *sql.DB
+	insertKey *sql.Stmt
+	countKeys *sql.Stmt
+}
+
+func (gc *GCDB) Open(home string) error {
+	db, err := sql.Open("sqlite3", "") // use temporary db
+	if err != nil {
+		return err
+	}
+	gc.db = db
+
+	_, err = db.Exec("CREATE TABLE Refs (key VARCHAR(64) PRIMARY KEY)")
+	if err != nil {
+		return err
+	}
+
+	insertKey, err := db.Prepare("INSERT INTO Refs VALUES (?)")
+	if err != nil {
+		return err
+	}
+	gc.insertKey = insertKey
+
+	countKeys, err := db.Prepare("SELECT COUNT(1) FROM Refs WHERE key = ?")
```

Comment:
Given that you're checking for existence (> 0) this should have a `LIMIT 1` to stop scanning after finding a match. I also wonder if `SELECT EXISTS(SELECT 1 FROM Refs WHERE key = ?)` is gonna get optimized to some nice index scan, maybe ask `EXPLAIN`?